src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,261 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <gsl/gsl_errno.h>
+#include <gsl/gsl_integration.h>
+#include <tuple>
+
+#include "Parallel/Printf.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace integrate {
+
+struct Result {
+  double value;
+  double error;
+};
+
+namespace detail {
+// forward decleration
+template <typename... Args>
+class GslQuadAdaptiveImpl;
+
+// ExpandTupleImpl takes and unpacks all args to pass them to the integrand,
+// the Algorithms in GSL Quad do not allow for a lambda function to be passed
+template <typename Args, typename ArgsIndices =
+                             std::make_index_sequence<tmpl::size<Args>::value>>
+struct ExpandTupleImpl;
+
+template <typename... Args, size_t... ArgsIndices>
+struct ExpandTupleImpl<tmpl::list<Args...>,
+                       std::index_sequence<ArgsIndices...>> {
+  template <typename FunctionType>
+  static double apply(FunctionType&& function, const double x,
+                      const std::tuple<Args...>& args) {
+    return (*function)(x, std::get<ArgsIndices>(args)...);
+  }
+};
+
+template <typename... Args>
+double integrand(double x, void* params) {
+  auto args = reinterpret_cast<GslQuadAdaptiveImpl<Args...>*>(params);
```

Comment:
This function needs a comment that explains the use of `reinterpret_cast`. You can add it above line 56 and remove the comment to `ExpandTupleImpl`. I suggest something like this:

> The GSL functions require the integrand to have this particular function signature. In particular, any extra parameters to the functions must be passed as a `void*` and re-interpreted appropriately. We interpret the pointer as an instance of `GslQuadAdaptiveImpl` and use it to retrieve the integrand function and its extra parameters.